File: src/state/machine.js
/**
 * Conversation State Machine
 * 
 * Manages conversation states and transitions.
 * Ensures deterministic flow through engagement phases.
 */

const STATES = {
  INITIAL: 'INITIAL',
  GREETING: 'GREETING',
  BUILDING_RAPPORT: 'BUILDING_RAPPORT',
  FINANCIAL_CONTEXT: 'FINANCIAL_CONTEXT',
  REQUEST: 'REQUEST',
  EXTRACTION: 'EXTRACTION',
  SUSPICIOUS: 'SUSPICIOUS',
  CLOSING: 'CLOSING',
  ENDED: 'ENDED'
};

const STATE_CONFIGS = {
  INITIAL: {
    next: ['GREETING'],
    maxTurns: 2,
    canEnd: false
  },
  GREETING: {
    next: ['BUILDING_RAPPORT'],
    maxTurns: 4,
    canEnd: false
  },
  BUILDING_RAPPORT: {
    next: ['FINANCIAL_CONTEXT', 'REQUEST'],
    maxTurns: 6,
    canEnd: false
  },
  FINANCIAL_CONTEXT: {
    next: ['REQUEST', 'EXTRACTION'],
    maxTurns: 8,
    canEnd: false
  },
  REQUEST: {
    next: ['EXTRACTION', 'FINANCIAL_CONTEXT', 'SUSPICIOUS'],
    maxTurns: 6,
    canEnd: false
  },
  EXTRACTION: {
    next: ['CLOSING', 'SUSPICIOUS'],
    maxTurns: 15,
    canEnd: false
  },
  SUSPICIOUS: {
    next: ['EXTRACTION', 'CLOSING', 'ENDED'],
    maxTurns: 5,
    canEnd: true
  },
  CLOSING: {
    next: ['ENDED'],
    maxTurns: 3,
    canEnd: true
  },
  ENDED: {
    next: [],
    maxTurns: 0,
    canEnd: true
  }
};

class StateMachine {
  constructor(config = {}) {
    this.states = STATES;
    this.configs = STATE_CONFIGS;
    this.maxTurns = config.maxTurns || 25;
    this.maxDuration = config.maxDuration || 30 * 60 * 1000; // 30 minutes
  }

  /**
   * Transition to next state based on context
   * @param {string} currentState - Current state
   * @param {Object} context - Transition context
   * @returns {Object} Transition result
   */
  transition(currentState, context) {
    const config = this.configs[currentState];
    
    // Check if already in terminal state
    if (currentState === STATES.ENDED) {
      return {
        current_state: STATES.ENDED,
        next_state: STATES.ENDED,
        reason: 'Already in terminal state',
        confidence: 1.0,
        should_end: true,
        termination_strategy: 'auto_close'
      };
    }

    // Check for forced transitions
    const forcedTransition = this.checkForcedTransitions(currentState, context);
    if (forcedTransition) {
      return forcedTransition;
    }

    // Check turn limits
    if (context.turnCount >= config.maxTurns) {
      return {
        current_state: currentState,
        next_state: STATES.CLOSING,
        reason: `Max turns (${config.maxTurns}) reached`,
        confidence: 0.9,
        should_end: false,
        termination_strategy: 'time_excuse'
      };
    }

    // State-specific transition logic
    switch (currentState) {
      case STATES.INITIAL:
        return this.transitionFromInitial(context);
      
      case STATES.GREETING:
        return this.transitionFromGreeting(context);
      
      case STATES.BUILDING_RAPPORT:
        return this.transitionFromBuildingRapport(context);
      
      case STATES.FINANCIAL_CONTEXT:
        return this.transitionFromFinancialContext(context);
      
      case STATES.REQUEST:
        return this.transitionFromRequest(context);
      
      case STATES.EXTRACTION:
        return this.transitionFromExtraction(context);
      
      case STATES.SUSPICIOUS:
        return this.transitionFromSuspicious(context);
      
      case STATES.CLOSING:
        return this.transitionFromClosing(context);
      
      default:
        return {
          current_state: currentState,
          next_state: currentState,
          reason: 'Unknown state, maintaining',
          confidence: 0.5,
          should_end: false
        };
    }
  }

  transitionFromInitial(context) {
    if (context.scamConfidence >= 0.75) {
      return {
        current_state: STATES.INITIAL,
        next_state: STATES.GREETING,
        reason: 'Scam confirmed (confidence >= 0.75)',
        confidence: 0.95,
        should_end: false
      };
    }
    
    return {
      current_state: STATES.INITIAL,
      next_state: STATES.GREETING,
      reason: 'Proceeding with caution',
      confidence: 0.6,
      should_end: false
    };
  }

  transitionFromGreeting(context) {
    if (context.turnCount >= 3 || context.hasFinancialContext) {
      return {
        current_state: STATES.GREETING,
        next_state: STATES.BUILDING_RAPPORT,
        reason: 'Greeting complete, building rapport',
        confidence: 0.85,
        should_end: false
      };
    }

    return {
      current_state: STATES.GREETING,
      next_state: STATES.GREETING,
      reason: 'Continuing greeting',
      confidence: 0.7,
      should_end: false
    };
  }

  transitionFromBuildingRapport(context) {
    if (context.turnCount >= 5 || context.hasDirectRequest) {
      return {
        current_state: STATES.BUILDING_RAPPORT,
        next_state: STATES.FINANCIAL_CONTEXT,
        reason: 'Rapport established, moving to financial context',
        confidence: 0.8,
        should_end: false
      };
    }

    if (context.hasFinancialContext) {
      return {
        current_state: STATES.BUILDING_RAPPORT,
        next_state: STATES.FINANCIAL_CONTEXT,
        reason: 'Financial context detected',
        confidence: 0.9,
        should_end: false
      };
    }

    return {
      current_state: STATES.BUILDING_RAPPORT,
      next_state: STATES.BUILDING_RAPPORT,
      reason: 'Continuing to build rapport',
      confidence: 0.6,
      should_end: false
    };
  }

  transitionFromFinancialContext(context) {
    if (context.turnCount >= 5 || context.hasDirectRequest) {
      return {
        current_state: STATES.FINANCIAL_CONTEXT,
        next_state: STATES.REQUEST,
        reason: 'Financial context understood, request received',
        confidence: 0.85,
        should_end: false
      };
    }

    return {
      current_state: STATES.FINANCIAL_CONTEXT,
      next_state: STATES.FINANCIAL_CONTEXT,
      reason: 'Clarifying financial details',
      confidence: 0.7,
      should_end: false
    };
  }

  transitionFromRequest(context) {
    if (context.turnCount >= 2) {
      return {
        current_state: STATES.REQUEST,
        next_state: STATES.EXTRACTION,
        reason: 'Request acknowledged, moving to extraction',
        confidence: 0.8,
        should_end: false
      };
    }

    return {
      current_state: STATES.REQUEST,
      next_state: STATES.REQUEST,
      reason: 'Clarifying request details',
      confidence: 0.6,
      should_end: false
    };
  }

  transitionFromExtraction(context) {
    // Only close if we have really good extraction OR very high turn count
    const extractionComplete = context.extractionProgress >= 0.9;
    
    // Much higher threshold before going to CLOSING
    if (extractionComplete && context.turnCount >= 12) {
      return {
        current_state: STATES.EXTRACTION,
        next_state: STATES.CLOSING,
        reason: 'Extraction complete with good data',
        confidence: 0.9,
        should_end: false,
        termination_strategy: 'graceful_closing'
      };
    }

    // Only timeout at very high turn count
    if (context.turnCount >= 15) {
      return {
        current_state: STATES.EXTRACTION,
        next_state: STATES.CLOSING,
        reason: 'Max extraction turns reached',
        confidence: 0.9,
        should_end: false,
        termination_strategy: 'graceful_closing'
      };
    }

    // Check for suspicion indicators
    if (context.consecutiveDelays >= 4) {
      return {
        current_state: STATES.EXTRACTION,
        next_state: STATES.SUSPICIOUS,
        reason: 'Too many delays, scammer may be suspicious',
        confidence: 0.7,
        should_end: false
      };
    }

    // Stay in EXTRACTION to keep engaging
    return {
      current_state: STATES.EXTRACTION,
      next_state: STATES.EXTRACTION,
      reason: 'Continuing extraction - more data needed',
      confidence: 0.7,
      should_end: false
    };
  }

  transitionFromSuspicious(context) {
    if (context.consecutiveDelays >= 3) {
      return {
        current_state: STATES.SUSPICIOUS,
        next_state: STATES.CLOSING,
        reason: 'Recovery failed, exiting gracefully',
        confidence: 0.8,
        should_end: false,
        termination_strategy: 'safe_exit'
      };
    }

    return {
      current_state: STATES.SUSPICIOUS,
      next_state: STATES.EXTRACTION,
      reason: 'Recovering from suspicion',
      confidence: 0.5,
      should_end: false
    };
  }

  transitionFromClosing(context) {
    return {
      current_state: STATES.CLOSING,
      next_state: STATES.ENDED,
      reason: 'Conversation complete',
      confidence: 1.0,
      should_end: true,
      termination_strategy: 'graceful_closing'
    };
  }

  checkForcedTransitions(currentState, context) {
    // Check for early termination conditions
    if (context.scammerTerminated) {
      return {
        current_state: currentState,
        next_state: STATES.ENDED,
        reason: 'Scammer terminated conversation',
        confidence: 1.0,
        should_end: true,
        termination_strategy: 'auto_close'
      };
    }

    return null;
  }

  /**
   * Get available transitions from current state
   * @param {string} state - Current state
   * @returns {string[]} Available next states
   */
  getAvailableTransitions(state) {
    return this.configs[state]?.next || [];
  }

  /**
   * Get state configuration
   * @param {string} state - State name
   * @returns {Object} State configuration
   */
  getStateConfig(state) {
    return this.configs[state] || null;
  }

  /**
   * Get all states
   * @returns {Object} All states
   */
  getAllStates() {
    return { ...this.states };
  }
}

module.exports = { StateMachine, STATES };

File: app/state/machine.py
"""
Conversation State Machine — Manages conversation lifecycle.

9 states: INITIAL → GREETING → BUILDING_RAPPORT → FINANCIAL_CONTEXT
          → REQUEST → EXTRACTION → SUSPICIOUS → CLOSING → ENDED
"""

from dataclasses import dataclass
from enum import Enum


class State(str, Enum):
    INITIAL = "INITIAL"
    GREETING = "GREETING"
    BUILDING_RAPPORT = "BUILDING_RAPPORT"
    FINANCIAL_CONTEXT = "FINANCIAL_CONTEXT"
    REQUEST = "REQUEST"
    EXTRACTION = "EXTRACTION"
    SUSPICIOUS = "SUSPICIOUS"
    CLOSING = "CLOSING"
    ENDED = "ENDED"


@dataclass
class TransitionResult:
    next_state: State
    should_end: bool = False
    reason: str = ""


STATE_CONFIGS = {
    State.INITIAL: {"max_turns": 1, "next": [State.GREETING]},
    State.GREETING: {"max_turns": 3, "next": [State.BUILDING_RAPPORT, State.FINANCIAL_CONTEXT]},
    State.BUILDING_RAPPORT: {"max_turns": 5, "next": [State.FINANCIAL_CONTEXT, State.REQUEST]},
    State.FINANCIAL_CONTEXT: {"max_turns": 5, "next": [State.REQUEST, State.EXTRACTION]},
    State.REQUEST: {"max_turns": 3, "next": [State.EXTRACTION, State.SUSPICIOUS]},
    State.EXTRACTION: {"max_turns": 15, "next": [State.CLOSING, State.SUSPICIOUS]},
    State.SUSPICIOUS: {"max_turns": 3, "next": [State.EXTRACTION, State.CLOSING, State.ENDED]},
    State.CLOSING: {"max_turns": 2, "next": [State.ENDED]},
    State.ENDED: {"max_turns": 0, "next": []},
}


class StateMachine:
    """Manage conversation state transitions."""

    def __init__(self):
        self.current_state = State.INITIAL
        self.turn_count = 0
        self.state_turn_count = 0
        self.history: list[dict] = []

    def transition(
        self,
        scam_confidence: float = 0.0,
        has_financial_context: bool = False,
        has_direct_request: bool = False,
        extraction_progress: float = 0.0,
        scammer_terminated: bool = False,
    ) -> TransitionResult:
        """Determine the next state based on context."""
        self.turn_count += 1
        self.state_turn_count += 1
        current = self.current_state
        config = STATE_CONFIGS[current]

        # Terminal state
        if current == State.ENDED:
            return TransitionResult(State.ENDED, should_end=True, reason="Conversation ended")

        # Scammer terminated
        if scammer_terminated:
            result = TransitionResult(State.ENDED, should_end=True, reason="Scammer terminated")
            self._do_transition(result.next_state, result.reason)
            return result

        next_state = current
        reason = "No transition needed"

        if current == State.INITIAL:
            next_state = State.GREETING
            reason = "Initial classification done"

        elif current == State.GREETING:
            if has_financial_context or has_direct_request:
                next_state = State.FINANCIAL_CONTEXT
                reason = "Financial context detected early"
            elif self.state_turn_count >= config["max_turns"]:
                next_state = State.BUILDING_RAPPORT
                reason = "Max greeting turns reached"

        elif current == State.BUILDING_RAPPORT:
            if has_direct_request:
                next_state = State.REQUEST
                reason = "Direct request received"
            elif has_financial_context or self.state_turn_count >= config["max_turns"]:
                next_state = State.FINANCIAL_CONTEXT
                reason = "Financial context detected or max turns"

        elif current == State.FINANCIAL_CONTEXT:
            if has_direct_request:
                next_state = State.REQUEST
                reason = "Direct request received"
            elif self.state_turn_count >= config["max_turns"]:
                next_state = State.REQUEST
                reason = "Max turns, moving to request"

        elif current == State.REQUEST:
            if self.state_turn_count >= config["max_turns"]:
                next_state = State.EXTRACTION
                reason = "Moving to extraction phase"

        elif current == State.EXTRACTION:
            if extraction_progress >= 0.9 and self.turn_count >= 12:
                next_state = State.CLOSING
                reason = "Extraction targets met"
            elif self.state_turn_count >= config["max_turns"]:
                next_state = State.CLOSING
                reason = "Max extraction turns reached"

        elif current == State.SUSPICIOUS:
            if self.state_turn_count >= config["max_turns"]:
                next_state = State.CLOSING
                reason = "Could not recover from suspicion"
            else:
                next_state = State.EXTRACTION
                reason = "Recovery attempt"

        elif current == State.CLOSING:
            if self.state_turn_count >= config["max_turns"]:
                next_state = State.ENDED
                reason = "Closing complete"

        if next_state != current:
            self._do_transition(next_state, reason)

        return TransitionResult(
            next_state=self.current_state,
            should_end=self.current_state == State.ENDED,
            reason=reason,
        )

    def _do_transition(self, new_state: State, reason: str):
        self.history.append({
            "from": self.current_state.value,
            "to": new_state.value,
            "turn": self.turn_count,
            "reason": reason,
        })
        self.current_state = new_state
        self.state_turn_count = 0

    def get_state(self) -> str:
        return self.current_state.value

    def to_dict(self) -> dict:
        return {
            "current_state": self.current_state.value,
            "turn_count": self.turn_count,
            "state_turn_count": self.state_turn_count,
            "history": self.history,
        }
